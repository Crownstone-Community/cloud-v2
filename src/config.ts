/**
 * This file will house all used environmental variables we use in the application.
 * It does not cover environmental variables used in dependencies.
 */
export const CONFIG = {
  mongoFilesURL: process.env.FILES_DB_URL ?? "mongodb://127.0.0.1:27017/files_v1",
  mongoDataURL:  process.env.DATA_DB_URL  ?? "mongodb://127.0.0.1:27017/data_v1",
  mongoUserURL:  process.env.USER_DB_URL  ?? "mongodb://127.0.0.1:27017/users_v1",

  emailValidationRequired: process.env.EMAIL_VALIDATION_REQUIRED ?? true,
  generateCustomIds: process.env.GENERATE_CUSTOM_IDS ?? false,

  SSEToken: process.env.SSE_TOKEN || "sse_token"
}

export const SecurityTypes = {
  accessToken: 'AccessToken',
}

