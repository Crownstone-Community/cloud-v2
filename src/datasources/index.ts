export * from './data.datasource';
export * from './users.datasource';
export * from './files.datasource';