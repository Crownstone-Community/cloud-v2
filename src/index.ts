import {ApplicationConfig, ExpressServer} from './server';
import {PopulateRepositoryContainer} from "./modules/containers/RepoContainer";

export * from './server';

const config = {
  rest: {
    // Use the LB4 application as a route. It should not be listening.
    listenOnStart: false,
  },
};

export async function main(options: ApplicationConfig = {}) {

  const server = new ExpressServer(config);
  await server.boot();
  await server.start();

  await PopulateRepositoryContainer(server.lbApp);

  const port = server.lbApp.restServer.config.port ?? 3000;
  const host = server.lbApp.restServer.config.host ?? 'NO-HOST';

  console.log(`Server is running at ${host}:${port}`);

  return server.lbApp;
}

if (require.main === module) {
  // Run the application
  main({}).catch(err => {
    console.error('Cannot start the application.', err);
    process.exit(1);
  });
}
