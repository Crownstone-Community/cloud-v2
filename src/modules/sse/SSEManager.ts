import { Server, Socket } from "socket.io"
import * as http from "http";
import Timeout = NodeJS.Timeout;
import * as crypto from "crypto";
import {Dbs} from "../containers/RepoContainer";
let config  = require('../config.' + (process.env.NODE_ENV || 'local'));

const S4 = function () {
  return Math.floor(Math.random() * 0x10000 /* 65536 */).toString(16);
};

const getShortUUID = function() {
  return (
    S4() + S4() + '-' +
    S4()
  );
}

const EVENT_ROOM_NAME = "/SSE_EVENTS";

const protocolTopics = {
  requestForOauthTokenCheck:  "requestForOauthTokenCheck",
  requestForAccessTokenCheck: "requestForAccessTokenCheck",
  authenticationRequest:      "authenticationRequest",
  event:                      "event",
}



class SSEManagerClass {
  io : Server;
  connections : {[id: string]: SSEConnection} = {};

  constructor() {
    this.io = null;
    this.connections = {};
  }

  init(server: http.Server) {
    this.io = new Server(server, { pingInterval: 4000, pingTimeout: 2000, transports:["websocket"], cookie:false })
    this.io.on('connect', (socket: Socket) => {
      let uid = getShortUUID()
      this.connections[uid] = new SSEConnection(socket, () => { delete this.connections[uid]; });
    });
  }

  emit(data: SseEvent) {
    this.io.sockets.in(EVENT_ROOM_NAME).emit(protocolTopics.event, data);
  }
}

export const SSEManager = new SSEManagerClass();


class SSEConnection {

  socket : Socket
  cleanup : () => void;
  authenticationTimeout : Timeout;
  handshake : string;

  constructor(socket: Socket, cleanup: () => void) {
    this.socket = null;
    this.cleanup = null;
    this.authenticationTimeout = null;
    this.handshake = null;

    this.handshake = crypto.randomBytes(16).toString('base64')

    this.socket = socket;
    this.cleanup = cleanup;

    // client gets 500 ms to authenticate
    this.authenticationTimeout = setTimeout(() => { this.destroy(); }, 500);

    // here we ensure only our SSE servers will connect to this socket before sending data.
    this.socket.emit(protocolTopics.authenticationRequest, this.handshake, (reply: string) => {
      if (this.authenticate(reply) === false) { this.destroy(); return; }
      clearTimeout(this.authenticationTimeout);

      // add to the list of sockets that can process SSE's
      this.socket.join(EVENT_ROOM_NAME);

      // check if an accesstoken is valid.
      this.socket.on(protocolTopics.requestForAccessTokenCheck, (token, callback) => {
        this.handleAccessTokenRequest(token, callback);
      })
      // check if an accesstoken is valid.
      this.socket.on(protocolTopics.requestForOauthTokenCheck, (token, callback) => {
        this.handleOauthTokenRequest(token, callback);
      })
    });

    this.socket.on("disconnect", () => { this.destroy(); });
  }


  destroy() {
    this.socket.disconnect(true);
    clearTimeout(this.authenticationTimeout);
    this.cleanup();
  }


  authenticate(reply: string) {
    let hasher = crypto.createHash('sha256');
    let output = hasher.update(this.handshake + config.SSEToken).digest('hex')
    return reply === output;
  }


  async _isValidAccessToken(token: string) : Promise<AccessModel> {
    let resultTokenData: AccessModel = {
      accessToken: token,
      ttl: 0,
      createdAt: new Date(),
      userId: null,
      spheres: {},
      scopes: ["all"],
    };

    // get the token from the db, if successful,
    let tokenData = await Dbs.crownstoneToken.findById(token);
    if (!tokenData) { throw "INVALID_TOKEN" }
    if (new Date().valueOf() < new Date(tokenData.created).valueOf() + tokenData.ttl*1000) {
      throw "EXPIRED_TOKEN"
    }
    resultTokenData.ttl = tokenData.ttl;
    resultTokenData.createdAt = tokenData.created;

    resultTokenData.userId = tokenData.userId;

    let access = await Dbs.sphereAccess.find({where: {and: [{userId: tokenData.userId}, {invitePending: {neq: true}}]}, fields: {sphereId: true}})

    for (let i = 0; i < access.length; i++) {
      resultTokenData.spheres[access[i].sphereId] = true;
    }
    return resultTokenData;
  }

  async _isValidOauthToken(token: string) : Promise<AccessModel> {
    let resultTokenData: AccessModel = {
      accessToken: token,
      ttl: 0,
      createdAt: new Date(),
      userId: null,
      spheres: {},
      scopes: [],
    };

    // get the token from the db, if successful,
    let tokenData = await Dbs.oauthToken.findById(token);
    if (!tokenData) { throw "INVALID_TOKEN" }
    if (new Date().valueOf() < new Date(tokenData.expiredAt).valueOf()) {
      throw "EXPIRED_TOKEN"
    }
    resultTokenData.ttl       = tokenData.expiresIn;
    resultTokenData.createdAt = tokenData.issuedAt;
    resultTokenData.userId    = tokenData.userId;
    resultTokenData.scopes    = tokenData.scopes as oauthScope[];

    let access = await Dbs.sphereAccess.find({where: {and: [{userId: tokenData.userId}, {invitePending: {neq: true}}]}, fields: {sphereId: true}})

    for (let i = 0; i < access.length; i++) {
      resultTokenData.spheres[access[i].sphereId] = true;
    }
    return resultTokenData;
  }

  handleAccessTokenRequest(token: string, callback: (result: any) => void) {
    this._isValidAccessToken(token)
      .then((result) => {
        if (!result) { throw "Invalid Token" }

        callback({code: 200, data: result});
      })
      .catch((err) => {
        return callback({code: 401, err: err});
      })
  }

  handleOauthTokenRequest(token: string, callback: (result: any) => void) {
    this._isValidOauthToken(token)
      .then((result) => {
        if (!result) { throw "Invalid Token" }

        callback({code: 200, data: result});
      })
      .catch((err) => {
        return callback({code: 401, err: err});
      })
  }
}
