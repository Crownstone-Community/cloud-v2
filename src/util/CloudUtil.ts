import * as crypto from "crypto";
const uuid = require('node-uuid');


export const CloudUtil = {
  createKey: function() : string {
    return crypto.randomBytes(16).toString('hex');
  },

  createIBeaconUUID() : string {
    return uuid.v4();
  },

  createToken: function() : string {
    return crypto.randomBytes(32).toString('hex');
  },

  hashPassword(plaintextPassword: string) : string {
    let shasum = crypto.createHash('sha1');
    shasum.update(String(plaintextPassword));
    let hashedPassword = shasum.digest('hex');
    return hashedPassword;
  }
}