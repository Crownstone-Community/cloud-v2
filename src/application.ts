import {BootMixin} from '@loopback/boot';
import {ApplicationConfig} from '@loopback/core';
import {
  RestExplorerBindings,
  RestExplorerComponent,
} from '@loopback/rest-explorer';
import {RepositoryMixin} from '@loopback/repository';
import {RestApplication} from '@loopback/rest';
import {ServiceMixin} from '@loopback/service-proxy';
import path from 'path';
import {CrownstoneSequence} from './sequence';
import {
  AuthenticationComponent,
  registerAuthenticationStrategy
} from '@loopback/authentication';
import {
  AuthorizationComponent,
} from '@loopback/authorization';
import {UserService} from './services';
import {AccessTokenStrategy} from "./security/authentication-strategies/access-token-strategy";
import {SecurityTypes} from "./config";
import {SphereAuthorizationComponent} from "./security/authorization-strategies/authorization";

export interface PackageInfo {
  name: string;
  version: string;
  description: string;
}
const pkg: PackageInfo = require('../package.json');

export class CrownstoneCloud extends BootMixin(ServiceMixin(RepositoryMixin(RestApplication))) {
  constructor(options: ApplicationConfig = {}) {
    let executionPath = __dirname;
    if (options.customPath !== undefined) { executionPath = options.customPath; }
    let customPort = process.env.PORT || 3050;
    if (options.rest && options.rest.port !== undefined) {
      customPort = options.rest.port;
    }

    let customHost = process.env.HOST || '127.0.0.1';
    if (options.rest && options.rest.host !== undefined) {
      customHost = options.rest.host;
    }

    super({...options, rest: { ...options.rest, port: customPort, host: customHost }})

    this.api({
      openapi: '3.0.0',
      info: {title: pkg.name, version: pkg.version},
      paths: {},
      components: {securitySchemes: {
        [SecurityTypes.accessToken]: {
          type: 'apiKey',
          in: 'header',
          name:'Authorization'
        },
      }},
      servers:  [{url: '/api'}],
      security: [{[SecurityTypes.accessToken]: []}],
    });


    this.setUpBindings();
    // Bind authentication component related elements
    this.component(AuthenticationComponent);
    this.component(AuthorizationComponent);
    this.component(SphereAuthorizationComponent);


    // authentication
    registerAuthenticationStrategy(this, AccessTokenStrategy);

    // Set up the custom sequence
    this.sequence(CrownstoneSequence);

    // Set up default home page
    this.static('/', path.join(executionPath, '../public'));

    // Customize @loopback/rest-explorer configuration here
    this.configure(RestExplorerBindings.COMPONENT).to({ path: '/explorer' });
    this.component(RestExplorerComponent);

    this.projectRoot = executionPath;

    // We define both here to allow the testing framework to generate coverage over the ts files.
    this.bootOptions = {
      controllers: {
        dirs: ['controllers'],
        extensions: ['.controller.ts','.controller.js'],
        nested: true,
      },
      repositories: {
        dirs: ['repositories'],
        extensions: ['.repository.ts','.repository.js'],
        nested: true,
      },
      datasources: {
        dirs: ['datasources'],
        extensions: ['.datasource.ts','.datasource.js'],
        nested: true,
      },
      services: {
        dirs: ['services'],
        extensions: ['.service.ts','.service.js'],
        nested: true,
      },
    };
  }

  setUpBindings(): void {
    this.bind("UserService").toClass(UserService);
  }
}
