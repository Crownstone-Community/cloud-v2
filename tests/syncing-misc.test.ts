import {CONFIG} from "../src/config";
CONFIG.emailValidationRequired = false;
CONFIG.generateCustomIds = true;

import {mockSSEmanager} from "./mocks/SSE";
mockSSEmanager();
import {SSEManager} from "../src/modules/sse/SSEManager";

import {makeUtilDeterministic, resetMockDatabaseIds, resetMockRandom} from "./mocks/CloudUtil.mock";
makeUtilDeterministic();

import {CrownstoneCloud} from "../src/application";
import {Client, createRestAppClient} from '@loopback/testlab';
import {clearTestDatabase, createApp, getRepositories} from "./helpers";
import {createHub, createLocation, createSphere, createStone, createUser, resetUsers} from "./builders/createUserData";
import {auth, getToken, setAuthToUser} from "./rest-helpers/rest.helpers";
import {getEncryptionKeys} from "../src/modules/sync/helpers/KeyUtil";
import {CloudUtil} from "../src/util/CloudUtil";
import { mocked } from 'ts-jest/utils'
import {SyncHandler} from "../src/modules/sync/SyncHandler";

let app    : CrownstoneCloud;
let client : Client;

let dbs;
let admin;
let member;
let guest;
let sphere;
let hub;
let stone,  behaviour,  ability,  abilityProperty;
let stone2, behaviour2, ability2, abilityProperty2;
let stone3, behaviour3, ability3, abilityProperty3;
let location;
let token;

async function populate() {
  // fill with a bit of data for sync
  dbs = getRepositories();
  admin    = await createUser('test@test.com', 'test', 0);
  member   = await createUser('member@test.com', 'test', 0);
  guest    = await createUser('guest@test.com', 'test', 0);
  sphere   = await createSphere(admin.id, 'mySphere', 0);
  await dbs.sphereAccess.create({sphereId: sphere.id, userId: member.id, role:'member', sphereAuthorizationToken: CloudUtil.createToken()});
  await dbs.sphereAccess.create({sphereId: sphere.id, userId: guest.id, role:'guest', sphereAuthorizationToken: CloudUtil.createToken()});
  hub      = await createHub(sphere.id, 'myHub', 0);
  ({stone, behaviour, ability, abilityProperty} = await createStone(sphere.id, 'stone1', 0));
  ({stone: stone2, behaviour: behaviour2, ability: ability2, abilityProperty: abilityProperty2} = await createStone(sphere.id, 'stone2', 0));
  ({stone: stone3, behaviour: behaviour3, ability: ability3, abilityProperty: abilityProperty3} = await createStone(sphere.id, 'stone3', 0));
  location = await createLocation(sphere.id, 'location', 0);

  stone.locationId = location.id;
  await dbs.stone.update(stone)
  token  = await getToken(client, admin);
}

beforeEach(async () => {
  mocked(SSEManager.emit).mockReset();
  await clearTestDatabase();
  resetUsers();
  resetMockRandom();
  resetMockDatabaseIds();
})
beforeAll(async () => {
  app    = await createApp()
  client = createRestAppClient(app);
})
afterAll(async () => { await app.stop(); })

test("get encryption keys", async () => {
  await populate();
  let adminKeys  = await getEncryptionKeys(admin.id);
  let memberKeys = await getEncryptionKeys(member.id);
  let guestKeys  = await getEncryptionKeys(guest.id);

  expect(adminKeys).toHaveLength(1);
  expect(adminKeys[0].sphereKeys).toHaveLength(7);
  expect(Object.keys(adminKeys[0].stoneKeys)).toHaveLength(3);
  expect(memberKeys).toHaveLength(1);
  expect(memberKeys[0].sphereKeys).toHaveLength(4);
  expect(Object.keys(memberKeys[0].stoneKeys)).toHaveLength(0);
  expect(guestKeys).toHaveLength(1);
  expect(guestKeys[0].sphereKeys).toHaveLength(3);
  expect(Object.keys(guestKeys[0].stoneKeys)).toHaveLength(0);
})


test("Sync FULL", async () => {
  await populate();
  await client.post(auth("/user/sync"))
    .expect(200)
    .send({sync: {type: "FULL"}})
    .expect(({body}) => {
      expect(body).toMatchSnapshot();
    })
});

test("Sync FULL with scope", async () => {
  await populate();
  let sphereId = sphere.id;


  await client.post(auth("/user/sync"))
    .send({sync: {type: "FULL", scope: ['spheres', 'hubs']}})
    .expect(({body}) => {
      expect(Object.keys(body)).toEqual(['spheres'])
      let sphere = body.spheres[sphereId];
      expect(Object.keys(sphere)).toEqual(['data', 'hubs'])
    })
})

test("Download users from sphere", async () => {
  await populate();
  let result = await SyncHandler.handleSync(admin.id, {sync: {type: "REQUEST"}, spheres: {[sphere.id]: {users:{member:{}, basic: {}, admin:{
    [admin.id]: {data: { updatedAt: admin.updatedAt},invitePending: false},
    ['hello']:  {data: { updatedAt: admin.updatedAt},invitePending: false},
  }}}}})

  expect(result.spheres[sphere.id].users.admin).toBeDefined()
  expect(result.spheres[sphere.id].users.admin[admin.id].data.status).toBe("IN_SYNC")
  expect(result.spheres[sphere.id].users.admin['hello'].data.status).toBe("NOT_AVAILABLE")
  expect(result.spheres[sphere.id].users.member[member.id].data.status).toBe("NEW_DATA_AVAILABLE")
})
