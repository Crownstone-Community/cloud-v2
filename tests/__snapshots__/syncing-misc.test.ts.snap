// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sync FULL 1`] = `
Object {
  "bootloaders": Object {
    "10102000100": null,
    "10102000200": null,
    "10102010000": null,
    "10102010300": null,
    "10103000100": null,
    "10103000200": null,
    "10103000300": null,
    "10103000400": null,
    "10103000500": null,
    "10103010000": null,
    "10104010000": null,
    "10105000000": null,
    "10106000000": null,
    "10106000100": null,
    "status": "VIEW",
  },
  "firmwares": Object {
    "10102000100": null,
    "10102000200": null,
    "10102010000": null,
    "10102010300": null,
    "10103000100": null,
    "10103000200": null,
    "10103000300": null,
    "10103000400": null,
    "10103000500": null,
    "10103010000": null,
    "10104010000": null,
    "10105000000": null,
    "10106000000": null,
    "10106000100": null,
    "status": "VIEW",
  },
  "keys": Array [
    Object {
      "sphereId": "dbId:SphereRepository:1",
      "sphereKeys": Array [
        Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "id": "dbId:SphereKeyRepository:1",
          "key": "key2005",
          "keyType": "ADMIN_KEY",
          "sphereId": "dbId:SphereRepository:1",
          "ttl": 0,
          "updatedAt": "1970-01-01T00:00:00.000Z",
        },
        Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "id": "dbId:SphereKeyRepository:2",
          "key": "key2006",
          "keyType": "MEMBER_KEY",
          "sphereId": "dbId:SphereRepository:1",
          "ttl": 0,
          "updatedAt": "1970-01-01T00:00:00.000Z",
        },
        Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "id": "dbId:SphereKeyRepository:3",
          "key": "key2007",
          "keyType": "BASIC_KEY",
          "sphereId": "dbId:SphereRepository:1",
          "ttl": 0,
          "updatedAt": "1970-01-01T00:00:00.000Z",
        },
        Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "id": "dbId:SphereKeyRepository:4",
          "key": "key2008",
          "keyType": "LOCALIZATION_KEY",
          "sphereId": "dbId:SphereRepository:1",
          "ttl": 0,
          "updatedAt": "1970-01-01T00:00:00.000Z",
        },
        Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "id": "dbId:SphereKeyRepository:5",
          "key": "key2009",
          "keyType": "SERVICE_DATA_KEY",
          "sphereId": "dbId:SphereRepository:1",
          "ttl": 0,
          "updatedAt": "1970-01-01T00:00:00.000Z",
        },
        Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "id": "dbId:SphereKeyRepository:6",
          "key": "key2010",
          "keyType": "MESH_APPLICATION_KEY",
          "sphereId": "dbId:SphereRepository:1",
          "ttl": 0,
          "updatedAt": "1970-01-01T00:00:00.000Z",
        },
        Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "id": "dbId:SphereKeyRepository:7",
          "key": "key2011",
          "keyType": "MESH_NETWORK_KEY",
          "sphereId": "dbId:SphereRepository:1",
          "ttl": 0,
          "updatedAt": "1970-01-01T00:00:00.000Z",
        },
      ],
      "stoneKeys": Object {
        "dbId:StoneRepository:1": Array [
          Object {
            "createdAt": "1970-01-01T00:00:00.000Z",
            "id": "dbId:StoneKeyRepository:1",
            "key": "key2019",
            "keyType": "UART_DEVICE_KEY",
            "sphereId": "dbId:SphereRepository:1",
            "stoneId": "dbId:StoneRepository:1",
            "ttl": 0,
            "updatedAt": "1970-01-01T00:00:00.000Z",
          },
          Object {
            "createdAt": "1970-01-01T00:00:00.000Z",
            "id": "dbId:StoneKeyRepository:2",
            "key": "key2020",
            "keyType": "MESH_DEVICE_KEY",
            "sphereId": "dbId:SphereRepository:1",
            "stoneId": "dbId:StoneRepository:1",
            "ttl": 0,
            "updatedAt": "1970-01-01T00:00:00.000Z",
          },
        ],
        "dbId:StoneRepository:2": Array [
          Object {
            "createdAt": "1970-01-01T00:00:00.000Z",
            "id": "dbId:StoneKeyRepository:3",
            "key": "key2024",
            "keyType": "UART_DEVICE_KEY",
            "sphereId": "dbId:SphereRepository:1",
            "stoneId": "dbId:StoneRepository:2",
            "ttl": 0,
            "updatedAt": "1970-01-01T00:00:00.000Z",
          },
          Object {
            "createdAt": "1970-01-01T00:00:00.000Z",
            "id": "dbId:StoneKeyRepository:4",
            "key": "key2025",
            "keyType": "MESH_DEVICE_KEY",
            "sphereId": "dbId:SphereRepository:1",
            "stoneId": "dbId:StoneRepository:2",
            "ttl": 0,
            "updatedAt": "1970-01-01T00:00:00.000Z",
          },
        ],
        "dbId:StoneRepository:3": Array [
          Object {
            "createdAt": "1970-01-01T00:00:00.000Z",
            "id": "dbId:StoneKeyRepository:5",
            "key": "key2029",
            "keyType": "UART_DEVICE_KEY",
            "sphereId": "dbId:SphereRepository:1",
            "stoneId": "dbId:StoneRepository:3",
            "ttl": 0,
            "updatedAt": "1970-01-01T00:00:00.000Z",
          },
          Object {
            "createdAt": "1970-01-01T00:00:00.000Z",
            "id": "dbId:StoneKeyRepository:6",
            "key": "key2030",
            "keyType": "MESH_DEVICE_KEY",
            "sphereId": "dbId:SphereRepository:1",
            "stoneId": "dbId:StoneRepository:3",
            "ttl": 0,
            "updatedAt": "1970-01-01T00:00:00.000Z",
          },
        ],
      },
    },
  ],
  "spheres": Object {
    "dbId:SphereRepository:1": Object {
      "data": Object {
        "data": Object {
          "createdAt": "1970-01-01T00:00:00.000Z",
          "id": "dbId:SphereRepository:1",
          "name": "mySphere",
          "uid": 212,
          "updatedAt": "1970-01-01T00:00:00.000Z",
          "uuid": "ibeaconUUID2003",
        },
        "status": "VIEW",
      },
      "hubs": Object {
        "dbId:HubRepository:1": Object {
          "data": Object {
            "data": Object {
              "createdAt": "1970-01-01T00:00:00.000Z",
              "httpPort": "80",
              "id": "dbId:HubRepository:1",
              "localIPAddress": "10.0.0.123",
              "name": "myHub",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
      },
      "locations": Object {
        "dbId:LocationRepository:1": Object {
          "data": Object {
            "data": Object {
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:LocationRepository:1",
              "name": "location",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
      },
      "stones": Object {
        "dbId:StoneRepository:1": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:1": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "enabled": false,
                  "id": "dbId:StoneAbilityRepository:1",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:1",
                  "syncedToCrownstone": false,
                  "type": "dimming",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:1": Object {
                  "data": Object {
                    "abilityId": "dbId:StoneAbilityRepository:1",
                    "createdAt": "1970-01-01T00:00:00.000Z",
                    "id": "dbId:StoneAbilityPropertyRepository:1",
                    "sphereId": "dbId:SphereRepository:1",
                    "stoneId": "dbId:StoneRepository:1",
                    "type": "smoothDimming",
                    "updatedAt": "1970-01-01T00:00:00.000Z",
                    "value": "true",
                  },
                  "status": "VIEW",
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:1": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "data": true,
                  "id": "dbId:StoneBehaviourRepository:1",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:1",
                  "syncedToCrownstone": false,
                  "type": "twilight",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
            },
          },
          "data": Object {
            "data": Object {
              "address": "RAN2016",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneRepository:1",
              "location": Object {
                "id": "dbId:LocationRepository:1",
                "name": "location",
              },
              "locationId": "dbId:LocationRepository:1",
              "major": 2017,
              "minor": 2018,
              "name": "stone1",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
        "dbId:StoneRepository:2": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:2": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "enabled": false,
                  "id": "dbId:StoneAbilityRepository:2",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:2",
                  "syncedToCrownstone": false,
                  "type": "dimming",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:2": Object {
                  "data": Object {
                    "abilityId": "dbId:StoneAbilityRepository:2",
                    "createdAt": "1970-01-01T00:00:00.000Z",
                    "id": "dbId:StoneAbilityPropertyRepository:2",
                    "sphereId": "dbId:SphereRepository:1",
                    "stoneId": "dbId:StoneRepository:2",
                    "type": "smoothDimming",
                    "updatedAt": "1970-01-01T00:00:00.000Z",
                    "value": "true",
                  },
                  "status": "VIEW",
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:2": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "data": true,
                  "id": "dbId:StoneBehaviourRepository:2",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:2",
                  "syncedToCrownstone": false,
                  "type": "twilight",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
            },
          },
          "data": Object {
            "data": Object {
              "address": "RAN2021",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneRepository:2",
              "major": 2022,
              "minor": 2023,
              "name": "stone2",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
        "dbId:StoneRepository:3": Object {
          "abilities": Object {
            "dbId:StoneAbilityRepository:3": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "enabled": false,
                  "id": "dbId:StoneAbilityRepository:3",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:3",
                  "syncedToCrownstone": false,
                  "type": "dimming",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
              "properties": Object {
                "dbId:StoneAbilityPropertyRepository:3": Object {
                  "data": Object {
                    "abilityId": "dbId:StoneAbilityRepository:3",
                    "createdAt": "1970-01-01T00:00:00.000Z",
                    "id": "dbId:StoneAbilityPropertyRepository:3",
                    "sphereId": "dbId:SphereRepository:1",
                    "stoneId": "dbId:StoneRepository:3",
                    "type": "smoothDimming",
                    "updatedAt": "1970-01-01T00:00:00.000Z",
                    "value": "true",
                  },
                  "status": "VIEW",
                },
              },
            },
          },
          "behaviours": Object {
            "dbId:StoneBehaviourRepository:3": Object {
              "data": Object {
                "data": Object {
                  "createdAt": "1970-01-01T00:00:00.000Z",
                  "data": true,
                  "id": "dbId:StoneBehaviourRepository:3",
                  "sphereId": "dbId:SphereRepository:1",
                  "stoneId": "dbId:StoneRepository:3",
                  "syncedToCrownstone": false,
                  "type": "twilight",
                  "updatedAt": "1970-01-01T00:00:00.000Z",
                },
                "status": "VIEW",
              },
            },
          },
          "data": Object {
            "data": Object {
              "address": "RAN2026",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "id": "dbId:StoneRepository:3",
              "major": 2027,
              "minor": 2028,
              "name": "stone3",
              "sphereId": "dbId:SphereRepository:1",
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
      },
      "users": Object {
        "dbId:UserRepository:1": Object {
          "data": Object {
            "data": Object {
              "accessLevel": "admin",
              "email": "test@test.com",
              "id": "dbId:UserRepository:1",
              "invitePending": false,
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
        "dbId:UserRepository:2": Object {
          "data": Object {
            "data": Object {
              "accessLevel": "member",
              "email": "member@test.com",
              "id": "dbId:UserRepository:2",
              "invitePending": false,
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
        "dbId:UserRepository:3": Object {
          "data": Object {
            "data": Object {
              "accessLevel": "guest",
              "email": "guest@test.com",
              "id": "dbId:UserRepository:3",
              "invitePending": false,
              "updatedAt": "1970-01-01T00:00:00.000Z",
            },
            "status": "VIEW",
          },
        },
      },
    },
  },
  "user": Object {
    "data": Object {
      "accountCreationPending": false,
      "createdAt": "1970-01-01T00:00:00.000Z",
      "email": "test@test.com",
      "emailVerified": false,
      "id": "dbId:UserRepository:1",
      "language": "en_us",
      "new": true,
      "updatedAt": "1970-01-01T00:00:00.000Z",
      "uploadDeviceDetails": true,
      "uploadLocation": true,
      "uploadSwitchState": true,
    },
    "status": "VIEW",
  },
}
`;
